src/map/guild.cpp

```diff
@@ -695,103 +695,150 @@ int guild_recv_info(struct guild *sg) {
  *--------------------------------------------*/
-int guild_invite(struct map_session_data *sd, struct map_session_data *tsd) {
-	struct guild *g;
-	int i;
+bool guild_invite( struct map_session_data *sd, struct map_session_data *tsd ){
+	nullpo_retr( false, sd );
 
-	nullpo_ret(sd);
+	// No nullpo_retr, because its valid that target players might not exist or are not online
+	if( tsd == nullptr ){
+		return false;
+	}
 
-	g=sd->guild;
+	struct guild *g = sd->guild;
 
-	if(tsd==NULL || g==NULL)
-		return 0;
+	nullpo_retr( false, g );
+
+	if( g->instance_id && battle_config.instance_block_invite ){
+		// TODO: message?
+		return false;
+	}
 
-	if( (i=guild_getposition(sd))<0 || !(g->position[i].mode&GUILD_PERM_INVITE) )
-		return 0; //Invite permission.
+	//Guild locked.
+	if( map_getmapflag( sd->bl.m, MF_GUILDLOCK ) ){
+		clif_displaymessage( sd->fd, msg_txt( sd, 228 ) );
+		return false;
+	}
+
+	// @noask [LuzZza]
+	if( tsd->state.noask ){
+		clif_noask_sub(sd, tsd, 2);
+		return false;
+	}
+
+	// Players in a clan can not join a guild
+	if( tsd->clan ){
+		return false;
+	}
+
+	// Invite permission.
+	if( !guild_has_permission( sd, GUILD_PERM_INVITE ) ){
+		return false;
+	}
 
 	if(!battle_config.invite_request_check) {
-	if (tsd->party_invite > 0 || tsd->trade_partner || tsd->adopt_invite) { //checking if there no other invitation pending
+		if (tsd->party_invite > 0 || tsd->trade_partner || tsd->adopt_invite) { //checking if there no other invitation pending
 			clif_guild_inviteack(sd,0);
-			return 0;
+			return false;
 		}
 	}
 
 	if (!tsd->fd) { //You can't invite someone who has already disconnected.
 		clif_guild_inviteack(sd,1);
-		return 0;
+		return false;
 	}
 
 	if(tsd->status.guild_id>0 ||
 		tsd->guild_invite>0 ||
 		map_flag_gvg2(tsd->bl.m))
 	{	//Can't invite people inside castles. [Skotlex]
 		clif_guild_inviteack(sd,0);
-		return 0;
+		return false;
 	}
 
+	int i;
+
 	//search an empty spot in guild
 	ARR_FIND( 0, g->max_member, i, g->member[i].account_id == 0 );
 	if(i==g->max_member){
 		clif_guild_inviteack(sd,3);
-		return 0;
+		return false;
 	}
 
 	tsd->guild_invite=sd->status.guild_id;
 	tsd->guild_invite_account=sd->status.account_id;
 
 	clif_guild_invite(tsd,g);
-	return 0;
+
+	return true;
 }
 
 /// Guild invitation reply.
 /// flag: 0:rejected, 1:accepted
-int guild_reply_invite(struct map_session_data* sd, int guild_id, int flag) {
-	struct map_session_data* tsd;
-
-	nullpo_ret(sd);
+bool guild_reply_invite( struct map_session_data* sd, int guild_id, int flag ){
+	nullpo_retr( false, sd );
 
 	// subsequent requests may override the value
-	if( sd->guild_invite != guild_id )
-		return 0; // mismatch
+	if( sd->guild_invite != guild_id ){
+		return false; // mismatch
+	}
 
 	// look up the person who sent the invite
 	//NOTE: this can be NULL because the person might have logged off in the meantime
-	tsd = map_id2sd(sd->guild_invite_account);
+	struct map_session_data *tsd = map_id2sd(sd->guild_invite_account);
```

Comment:
can this lookup be moved after "already in another guild" check ?
Since the another guild is not dependant.